### PR TITLE
Suggested Refactoring Plan for Breaking up the `MainUserSession` Monolith.

### PR DESCRIPTION
**Pull request by zzcgumn**
Adds a document describing a possible plan for refactoring `MainUserSession` from a monolith to a composite. It is, of course, possible to do this in several ways and everyone is encouraged to contribute suggestions.

I have not opened this PR as a proposal as it does not contain suggested implementations. It is intended to share what I have learned while inventing our technical debt, going through existing documentation and talking with other team members.

### CODE REVIEW

File: TechnicalDocuments/RefactorAppMainUserSession.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+More background information about what `MainUserSession` is and what we would like it to become can be found in this [deck](https://drive.google.com/drive/u/0/folders/1PVm5ztVYGpOdqLZxTmIl2BXff8ux8MnK), it is a Keynote presentation so you need download and unzip it.
+
+Proposed refactoring is to change AppMainUserSession from being a singular to a composite entity. It will still comply with the current set of protocols, but implement them by forwarding to the relevant component. This can, hopefully, be done in a step by step manner to avoid a high risk gargantuan pull request.
+
+# Step One, Move to the Babylon Main Project.
+
+Both the SessionViewModel, MainUserSession and AppMainUserSession are specific to the Babylon app targets and not needed to implement the SDK. They should thus be moved out of BabylonCore. To move them to the Octopus folder in the main project seems logical, but BabylonDependencies is also technically possible. Moving them to the main project will require minor refactoring to a couple of UI builders that consume MainUserSession directly.
```
Comment (andersio) [inline TechnicalDocuments/RefactorAppMainUserSession.md]: We should also move the UserSession protocols to BabylonDeps. They are not used by the SDK at all.

Comment (andersio) [reply TechnicalDocuments/RefactorAppMainUserSession.md]: It makes sense for the implementation to go to the main app target, btw.

File: TechnicalDocuments/RefactorAppMainUserSession.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+# Step Three, Add a `ClinicalRecords` Instance.
+
+Although it is technically possible to create `ClinicalRecords` instances on demand this needs be avoided since they hold a stateful authentication for accessing sensitive content. Creating them on demand without sharing the authentication status could result in the user being asked to supply their password or finger-print several times in the same app session.
+
+Services for accessing family accounts, biometric authentication and perform extra authentication provided by `MainUserSession` should be routed through the `ClinicalRecords` instance. This is unlikely to be a one to one match so will involve adapting builders and view models that deal with sensitive data.
+
+# Further Steps.
+
+Replace business controllers for the current user, id verification status and image handling with their SDK counterparts.
```
Comment (viorel15) [inline TechnicalDocuments/RefactorAppMainUserSession.md]: MainUser session functionality should be gradually replaced with the existing SDK's

File: TechnicalDocuments/RefactorAppMainUserSession.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+# Step Two, Add an `AuthenticatedContext` Instance.
+
+Serving an `AuthenticatedContext` through the `MainUserSession` interface will allow builders to create instances of all SDK APIs. It should be noted that all our SDK APIs are very lightweight both in terms of memory consumption and time to create. There is thus no need to serve them centrally unless they hold state that should be shared.
+
+Methods in `MainUserSession` that implement requirements for `AuthenticatedAccessible` will be changed to route through the `AuthenticatedContext`.
+
+The `AuthenticatedContext` can be created by the `SessionViewModel` until sign in, sign up and the app boot strap has been refactored to employ the `Auth` SDK.
+
+# Step Three, Add a `ClinicalRecords` Instance.
```
Comment (viorel15) [inline TechnicalDocuments/RefactorAppMainUserSession.md]: Before this step I see another step"
# Replace `session` from SessionViewModel and from AppMainUSerSession with Auth instance.

First the SessionViewModel should use the `Auth.login` methods and and use the Authenticated context provided through that flow.

File: TechnicalDocuments/RefactorAppMainUserSession.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+# Step One, Move to the Babylon Main Project.
+
+Both the SessionViewModel, MainUserSession and AppMainUserSession are specific to the Babylon app targets and not needed to implement the SDK. They should thus be moved out of BabylonCore. To move them to the Octopus folder in the main project seems logical, but BabylonDependencies is also technically possible. Moving them to the main project will require minor refactoring to a couple of UI builders that consume MainUserSession directly.
+
+# Step Two, Add an `AuthenticatedContext` Instance.
+
+Serving an `AuthenticatedContext` through the `MainUserSession` interface will allow builders to create instances of all SDK APIs. It should be noted that all our SDK APIs are very lightweight both in terms of memory consumption and time to create. There is thus no need to serve them centrally unless they hold state that should be shared.
+
```
[the remark this answers is not in the document]
Comment (zzcgumn) [reply TechnicalDocuments/RefactorAppMainUserSession.md]: Maybe best to do this between my proposes `Step One` and `Step Two`.

File: TechnicalDocuments/RefactorAppMainUserSession.md (hunk 3 above)
Comment (zzcgumn) [reply TechnicalDocuments/RefactorAppMainUserSession.md]: Can we do this without refactoring sign in and sign up?

**Comment by diegopetrucci** [comment]
What is the current status of this proposal? Should we close it for the time being or are there plans to work on it?

**Comment by zzcgumn** [comment]
@diegopetrucci I think the status is work in progress. @viorel15 is working on updating our codebase to consume and authenticated context instead of Main User Session as a first step towards consuming the SDKs instead of business controllers directly.

**Comment by RuiAAPeres** [comment]
@zzcgumn any update on this? Otherwise, I will close it. ❤️

**Comment by zzcgumn** [comment]
Please check with @viorel15 if this write-up is still helpful. Parts of what is discussed here is certainly out of date now with the refactoring of session view model.

**Comment by RuiAAPeres** [comment]
Please re-open if this is still relevant. Thanks! ❤️

**Comment by zzcgumn** [comment]
I think this document has served its purpose.

**Pull request closed without merging**